### ISSUE TITLE
Getting not implemented trait when having nested enums/CHOICE

### ISSUE THREAD

**Issue by mtodescato** [alex/rust-asn1#317]
Hello basically if i have a struct which contains an enum everything works:
``#[derive(asn1::Asn1Read, asn1::Asn1Write)]
pub struct InfoAccessItem<'a> {
    oid: asn1::ObjectIdentifier,
    accessLocation: GeneralName<'a>,
}

#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
pub enum GeneralName<'a> {
    #[implicit(0)]
    OtherName(asn1::IA5String<'a>),
    #[implicit(1)]
    Rfc822Name(asn1::IA5String<'a>),
    #[implicit(2)]
    DNSName(asn1::IA5String<'a>),
    #[implicit(3)]
    X400Address(asn1::IA5String<'a>),
    #[implicit(4)]
    DirectoryName(asn1::IA5String<'a>),
    #[implicit(5)]
    EdiPartyName(asn1::IA5String<'a>),
    #[implicit(6)]
    UniformResourceIdentifier(asn1::IA5String<'a>),
    #[implicit(7)]
    IPAddress(&'a [u8]),
    #[implicit(8)]
    RegisteredID(asn1::ObjectIdentifier),
}``

If i Have an Enum which contain another enum like this:

``#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
pub struct DistributionPoint<'a> {
    #[implicit(0)]
    distributionPoint: Option<DistributionPointName<'a>>,
    #[implicit(1)]
    reasons: Option<asn1::BitString<'a>>,
    #[implicit(2)]
    crlissuer: Option<GeneralName<'a>>,
}

#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
enum DistributionPointName<'a> {
    #[implicit(0)]
    FullName(GeneralName<'a>),
    #[implicit(1)]
    NameRelativeToCRLIssuer(RelativeDistinguishedName<'a>),
}

#[derive(asn1::Asn1Read, asn1::Asn1Write)]
pub struct RelativeDistinguishedName<'a> {
    data: asn1::SequenceOf<'a, AttributeValueAssertion<'a>>,
}

#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
pub struct AttributeValueAssertion<'a> {
    attribute_type: asn1::ObjectIdentifier,
    attribute_value: asn1::IA5String<'a>,
}``

Where General name is the same as the previous example, it gives me an error every nested enum as write and read asn1 trait not implemented. 
Am I doing something wrong or is it a bug?

**Comment by alex** [comment]
Your problem is that the Full name choice of Distribution point name should
be a sequence of general name.

בתאריך יום ג׳, 30 באוג׳ 2022, 9:26, מאת matteotodescato ‏<
***@***.***>:

**Comment by mtodescato** [comment]
In fact putting sequence of for general name and distribution point (needs to be a sequence of no idea why).
But I am wondering why it gives error, because i should be able to put them without the sequence of.
And for the distribution point field here:
```#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
pub struct DistributionPoint<'a> {
#[implicit(0)]
distributionPoint: Option<DistributionPointName<'a>>,
#[implicit(1)]
reasons: Option<asn1::BitString<'a>>,
#[implicit(2)]
crlissuer: Option<GeneralName<'a>>,
}` 
I should not be forced to make it a sequence of since the file i need to parse dosen't have a sequence of.

**Comment by alex** [comment]
The structures you have here appear to be from RFC5280, so when I say there should be a `SEQUENCE OF`, I mean that the RFC contains one, and I assumed you were trying to implement it.

In either event, the key limitation here is an inability to mark an `enum` as `#[implicit]`. This is due to a combination of ASN.1 semantics + rust-asn1 APIs:

In ASN.1, `IMPLICIT` fields have a tag which is the value specified, with the constructed bit set if the _inner_ type is constructed. This means to compute an implicit tag you need to know the inner type. Because an enum (obviously) has multiple types there's no one answer to "is the constructed bit set". Further, it's not clear how you'd know which enum element you had, since there'd only be two potential tags: with constructed bit and without, not one per element.

In conclusion: I don't think having a `CHOICE` field marked as `IMPLICIT` will be easy to support, or even really makes sense. Please let me know if you think I missed something in this analysis.

**Comment by mtodescato** [comment]
Yes I am implementing a structure which is RFC5280 because I am trying to parse something encoded in that way.
Maybe I am wrong, never really worked with asn1, but i made an enum marked implicit:
`#[derive(Debug, asn1::Asn1Read, asn1::Asn1Write)]
pub enum GeneralName<'a> {
    #[implicit(0)]
    OtherName(asn1::IA5String<'a>),
    #[implicit(1)]
    Rfc822Name(asn1::IA5String<'a>),
    #[implicit(2)]
    DNSName(asn1::IA5String<'a>),
    #[implicit(3)]
    X400Address(asn1::IA5String<'a>),
    #[implicit(4)]
    DirectoryName(asn1::IA5String<'a>),
    #[implicit(5)]
    EdiPartyName(asn1::IA5String<'a>),
    #[implicit(6)]
    UniformResourceIdentifier(asn1::IA5String<'a>),
    #[implicit(7)]
    IPAddress(&'a [u8]),
    #[implicit(8)]
    RegisteredID(asn1::ObjectIdentifier),
}`
I have already used it and it works, that's why I am confused.
And after the General name error which was my fault, but as i see it should work anyway, i get forced by the api to put a sequence of in the following code to make it compile:
`#[derive(asn1::Asn1Read, asn1::Asn1Write)]
pub struct DistributionPoint<'a> {
    #[implicit(0)]
    distributionPoint: Option<DistributionPointName<'a>>,
    #[implicit(1)]
    reasons: Option<asn1::BitString<'a>>,
    #[implicit(2)]
    crlissuer: Option<asn1::SequenceOf<'a, GeneralName<'a>>>,
}

#[derive(asn1::Asn1Read, asn1::Asn1Write)]
enum DistributionPointName<'a> {
    #[implicit(0)]
    FullName(asn1::SequenceOf<'a, GeneralName<'a>>),
    #[implicit(1)]
    NameRelativeToCRLIssuer(RelativeDistinguishedName<'a>),
}`

And on distributionPoint: Option<DistributionPointName<'a>>, it gives me the not implemented trait too.
Putting sequence of would make it compile, but i don't think it will parse since in the structure i need to pare that is not a list.

**Comment by alex** [comment]
FYI your messages are basically unreadable because you aren't using Github's markdown correctly. Please try to use the preview functionality and ensure messages are formatted.

Can you post a minimal example with error message, I'm having trouble following what you are saying doesn't work.

FWIW we've got a full implementation of X.509 with rust-asn1 in pyca/cryptography that you might find an interesting reference: https://github.com/pyca/cryptography/blob/main/src/rust/src/x509/certificate.rs#L613

**Comment by mtodescato** [comment]
Sorry for the markdown, I don't really use it
With the two following strucutre, which are the same you implemented in the link you sent
https://pastebin.com/enw1JNfn
I get the following error:
https://pastebin.com/b4dbccas

The implementation of x509 is available as a crate?

**Comment by alex** [comment]
You've got:

```
    #[implicit(0)]
    distributionPoint: Option<DistributionPointName<'a>>,
```

but it should be `explicit(0)`. From the RFC:

```
   DistributionPoint ::= SEQUENCE {
        distributionPoint       [0]     DistributionPointName OPTIONAL,
        reasons                 [1]     ReasonFlags OPTIONAL,
        cRLIssuer               [2]     GeneralNames OPTIONAL }
```

There's some rule in ASN.1 about the type of tagging changing from `implicit` to `explicit` and unfortunately I don't have exact citation for you :-(

Sorry ASN.1 is such a minefield.

**Comment by mtodescato** [comment]
Ok, I got it, the error of the compiler is really not clear.
Thanks for the help

**Comment by mtodescato** [comment]
Well I think the problem lies in the fact that you need to use a macro. But at the same time you don't really have a way to make the compiler know that it should be looking for implicit instead of explicit or viceversa, or at least I can't see an easy way of doing this.
I think now at least someone may find the issue and see he has the same error.

**Comment by alex** [comment]
I'm glad you were able to solve this, and hopefully this is a good resource for someone in the future. Going to close this as resolved.
